src/plugins/data/public/search/search_source/search_source.ts

```diff
@@ -381,6 +381,8 @@ export class SearchSource {
       case 'fields':
         const fields = uniq((data[key] || []).concat(val));
         return addToRoot(key, fields);
+      case 'fieldsApi':
```

Comment:
I think we need to start this as a larger discussion with App Arch and Stacey (feel free to ping them here once you want to open this PR for review). I am a bit torn on already having a `fields` API in the search source and now add a `fieldsApi` parameter to it, that will set this to the `fields` part of the query. I think the following code as a consumer is just not really understandable:

```ts
searchSource
  .setField('fields', ['bar', 'baz'])
  .setField('fieldsApi', ['some_field']);
```

Since `SearchSource` is a rather present API, we're needing in a lot of places, I'd not want to dilute it for a work-around that we need in Discover.

The `fields` field in search source, already does multiple things, like `stored_fields` and `docvalue_fields`. I think we should rather make sure that key will also handle `fields` (in the ES query) correctly. Maybe we can even replace `stored_fields` and `docvalue_fields` in that API completely by using the `fields` API for them too.

tl;dr: I think we need to have some discussion with App Arch to find a proper longer term solution for this part, the other part looks as the initial workaround fine for me.

Reply:
dear @elastic/kibana-app-arch & @stacey-gammon 

so the requirements are, that `SearchSource` would would provide access to the `search-fields` API

https://www.elastic.co/guide/en/elasticsearch/reference/master/search-fields.html

Like Tim mentioned the `fields` field was already taken, modifying it's functionality broke several functional tests, so I've added a `fieldsApi` field, which would provide the functionality `Discover` would need. Maybe it would make sense to create a separate issue for this, unless everybody thinks the solution provided in this PR is fine. I also would prefer a single `fields` field in SearchSource, but the approach I've tried to implement here broke the following functional tests:
https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/73366/

Reply:
Here's the issue: #77241

Reply:
thanks @kertal, we'll discuss this and get back to you. is this a blocker for discover at the moment ?

Reply:
hmm, that issue seems very general, should we open one for discover specifically ? what exactly is the issue for discover, the request searchsource sends out or the response provided ? searchsource will return full response so its up to you how you parse it.

Reply:
@ppisljar I've created an issue #77300 but closed in favor of #77241. It's basically the question: how to extend `SearchSource` in a non-hacky way to have access to the fields API of search. Without the extension runtime fields don't work in Discover.

Reply:
https://github.com/elastic/kibana/issues/77107 is the Discover specific issue that this PR attempts to fix.